honeycombio/honeycomb-network-agent
# feat: add option to skip options validation

## Which problem is this PR solving?
When setting an alternative API endpoint, API key validation will fail and prevents the agent from starting. 

This PR adds a config option that will skip config validation if desired. 

- Closes #264 

## Short description of the changes

- Add configuration option `HONEYCOMB_SKIP_OPTIONS_VALIDATION` that, when set to True, skips the Validate function that today primarily does a lookup on the API Key in use.
- Move the Config Validate below the logging of config values; that way, if there is something wrong, the config values are printed out first for easier troubleshooting if it fails.

## How to verify that this has the expected result

Set a different `HONEYCOMB_API_ENDPOINT`.

With `HONEYCOMB_SKIP_OPTIONS_VALIDATION=false` (default), pod will fail to start and show an error in the logs after showing the provided configuration options.

```sh
2023-10-02 16:30:43 {"level":"fatal","error":"Invalid API key","time":"2023-10-02T20:30:43Z","message":"Config validation failed"}.
```
With `HONEYCOMB_SKIP_OPTIONS_VALIDATION=true`, no error will throw and agent will start up as expected. Data may not be sent if API Key doesn't match the endpoint but the pod will not get stuck in backoff.

```yaml
# deployment.yaml
...
            ## uncomment this to set the destination dataset for events
            - name: HONEYCOMB_API_ENDPOINT
              value: $HONEYCOMB_API_ENDPOINT
            ## uncomment this to skip options validation
            - name: HONEYCOMB_SKIP_OPTIONS_VALIDATION
              value: "$HONEYCOMB_SKIP_OPTIONS_VALIDATION"
```
